### PR TITLE
Improve test utility header compilation speed

### PR DESCRIPTION
**Pull request by vittorioromeo**
This PR reduces physical dependencies between headers for test utilities, resulting in 1-2s average compilation speedup on my machine.

### CODE REVIEW

**Comment by vittorioromeo** [comment]
Failures seem to be related to codecov -- how to proceed?

**Comment by ChrisThrasher** [comment]
`master` (without running tests)
```
real    0m7.925s
user    0m56.337s
sys     0m6.599s
```

This PR (without running tests)
```
real    0m7.952s
user    0m56.119s
sys     0m6.551s
```

I'm not able to observe a build time reduction with this PR.

**Comment by ChrisThrasher** [comment]
> CI failures seem to be related to codecov -- how to proceed?

It fails all the time. Rerunning CI usually fixes it.

**Review by ChrisThrasher — CHANGES_REQUESTED** [review]

**Review by Bromeon — COMMENTED** [review]
Looks mostly good -- I wonder if we should still define the vector typedefs, as we're using them in a few places? Just for consistency with the rest of the codebase.

File: test/TestUtilities/GraphicsUtil.cpp [hunk 1]
```diff
@@ -1,8 +1,10 @@
 #include <SFML/Graphics/BlendMode.hpp>
 #include <SFML/Graphics/Color.hpp>
+#include <SFML/Graphics/Rect.hpp>
 #include <SFML/Graphics/Transform.hpp>
 
 #include <GraphicsUtil.hpp>
+#include <SystemUtil.hpp>
```
[the remark this answers is not in the document]
Comment (vittorioromeo) [reply test/TestUtilities/GraphicsUtil.cpp:7]: This is necessary as a transitive include, it was also necessary prior to this PR IIRC.

**Comment by vittorioromeo** [comment]
> Looks mostly good -- I wonder if we should still define the vector typedefs, as we're using them in a few places? Just for consistency with the rest of the codebase.

You mean in `SystemUtil`, right? I'd rather avoid doing so, it's a bit of extra repetition that we need to maintain between the actual vector header and the test header.

**Comment by vittorioromeo** [comment]
SFMLBuildMaster: Build this please.

File: test/TestUtilities/GraphicsUtil.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (ChrisThrasher) [reply test/TestUtilities/GraphicsUtil.cpp:7]: Necessary as in the build fails without this include?

**Review by ChrisThrasher — COMMENTED** [review]
I'm failing to demonstrate any compile time improvements and even if I could show that it was faster for me, I'm still not keen on adding complexity for marginal compile time gains that aren't even relevant to the core library. I won't veto this PR if others like it but I'm still not excited by it.

**Comment by vittorioromeo** [comment]
> I'm failing to demonstrate any compile time improvements and even if I could show that it was faster for me, I'm still not keen on adding complexity for marginal compile time gains that aren't even relevant to the core library. I won't veto this PR if others like it but I'm still not excited by it.

Here are some measurements on my machine, using `hyperfine "ninja clean; ninja"`, for a clean rebuild including all test drivers.

Before this PR:

```
  Time (mean ± σ):     15.025 s ±  0.531 s    [User: 120.985 s, System: 17.608 s]
  Range (min … max):   14.178 s … 15.642 s    10 runs
```

After this PR:

```
  Time (mean ± σ):     13.684 s ±  0.111 s    [User: 111.092 s, System: 16.656 s]
  Range (min … max):   13.537 s … 13.891 s    10 runs
```

**Comment by ChrisThrasher** [comment]
I was unaware of hyperfine. Looks pretty useful for these things. I again commented out the command that runs the tests since I don't want those skewing the results. (I'm growing sour on SFML running tests automatically after the build and might prefer we switch to he more idiomatic approach where tests are only run on demand. If we care about incremental build time improvements that's an easy way to improve).

`master`:
```
Benchmark 1: ninja clean; ninja
  Time (mean ± σ):      9.657 s ±  0.209 s    [User: 62.755 s, System: 7.995 s]
  Range (min … max):    9.424 s …  9.974 s    10 runs
```

This PR:
```
Benchmark 1: ninja clean; ninja
  Time (mean ± σ):      9.423 s ±  0.115 s    [User: 62.471 s, System: 7.878 s]
  Range (min … max):    9.303 s …  9.722 s    10 runs
```

The difference is still imperceptible to me.

**Comment by ChrisThrasher** [comment]
> CI failures seem to be related to codecov -- how to proceed?

https://github.com/ChrisThrasher/SFML/commit/5f8b7bf4f0c40fe039d6bf08836f446b4815da3f

Unrelated to this PR but I propose we allow coverage jobs to fail. I'd rather have a rock-solid CI pipeline than perfect code coverage information.

**Review by ChrisThrasher — COMMENTED** [review]
I think this increase in complexity is not worth it for the measured build time differences. Our tests need not have hyper-optimized compile times. I'd much rather keep them straightforward and easy to read. It's a "no" from me I'm afraid.

**Review by Bromeon — APPROVED** [review]
The added complexity is mostly about the explicit template instantiations, or am I missing something else?

I consider the changes in readability acceptable here, given that they're quite isolated and the improvements are measurable, even if small. I have more issues with changes that affect the entire codebase for arguable gains (e.g. large-scale code style changes).

**Comment by vittorioromeo** [comment]
Rebased on `master`. @ChrisThrasher, are you still against this PR given Bromeon's observation? We are very likely to not touch this code again, but it will continuously give us benefits in terms of compilation speed and CI usage over the course of the next years.

**Review by ChrisThrasher — COMMENTED** [review]
> Rebased on master. @ChrisThrasher, are you still against this PR given Bromeon's observation? We are very likely to not touch this code again, but it will continuously give us benefits in terms of compilation speed and CI usage over the course of the next years.

Sorry I'm still a "no", but I won't veto it. It ultimately comes down to how I value clear, simple code more than you and you value fast build times much more than me. We both value both things, but we value them to different degrees which leads us to making different conclusions.

I kindly ask that @eXpl0it3r step in and decide what SFML's philosophy is for optimizating compile times versus the readability and complexity tradeoffs that entails so that we can stop having these debates.

File: test/TestUtilities/GraphicsUtil.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (eXpl0it3r) [reply test/TestUtilities/GraphicsUtil.cpp:7]: It's certainly needed now, since the SystemUtil header defines `setStreamPrecision`

**Review by eXpl0it3r — APPROVED** [review]
Moved around a bit much for my taste to get a tiny improvement on your machine ™️ at the same time, I think it's not that much more complex.

We should however not lose sight of what we're trying to achieve here.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Comment by vittorioromeo** [comment]
@eXpl0it3r @ChrisThrasher: how do we proceed?

BTW, here's the latest benchmark after rebasing:

```
without PR:
Benchmark 1: ninja clean && ninja
  Time (mean ± σ):     16.227 s ±  0.570 s    [User: 169.491 s, System: 22.362 s]
  Range (min … max):   15.517 s … 17.353 s    10 runs

with PR:
Benchmark 1: ninja clean && ninja
  Time (mean ± σ):     15.539 s ±  0.481 s    [User: 161.124 s, System: 21.582 s]
  Range (min … max):   14.774 s … 16.092 s    10 runs
```

**Comment by ChrisThrasher** [comment]
I stand by my previous comments. In case anything changed I remeasured and the build time improvement was about a tenth of a second. It's up to Exploiter to decide.

**Comment by vittorioromeo** [comment]
> I stand by my previous comments. In case anything changed I remeasured and the build time improvement was about a tenth of a second. It's up to Exploiter to decide.

He approved the PR, but merging is blocked due to your review.

**Review by ChrisThrasher — COMMENTED** [review]
See previous comments

**Pull request merged**